server/rest/src/main/java/org/infinispan/rest/resources/ProtobufResource.java

```diff
@@ -0,0 +1,227 @@
+
+      AdvancedCache<Object, Object> cache = invocationHelper.getRestCacheManager()
+            .getCache(ProtobufMetadataManager.PROTOBUF_METADATA_CACHE_NAME, request);
+
+      return CompletableFuture.supplyAsync(() ->
+          Flowable.fromIterable(cache.keySet())
+               .filter(key -> !((String)key).endsWith(ProtobufMetadataManager.ERRORS_KEY_SUFFIX))
+               .map(key -> {
+                  String error = (String) cache
+                        .get(key + ProtobufMetadataManagerConstants.ERRORS_KEY_SUFFIX);
+                  ProtoSchema protoSchema = new ProtoSchema();
+                  protoSchema.name = (String) key;
+                  if (error != null) {
+                     protoSchema.error = createErrorContent(protoSchema.name, error);
+                  }
+                  return protoSchema;
+               })
+                .sorted(Comparator.comparing(s -> s.name))
+                .collect(Collectors.toList())
+                .map(protoSchemas -> {
+                  try {
+                     byte[] bytes = objectMapper.writeValueAsBytes(protoSchemas);
+                     responseBuilder.contentType(APPLICATION_JSON).entity(bytes);
+                  } catch (JsonProcessingException e) {
+                     throw new RestResponseException(e);
+                  }
+                  return (RestResponse) responseBuilder.build();
+               })
+                .toCompletionStage()
+      , invocationHelper.getExecutor())
+            .thenCompose(Function.identity());
+   }
+
+   private CompletionStage<RestResponse> addOrEditSchema(RestRequest request, boolean create) {
+
+      String schemaName = checkMandatorySchemaName(request);
+
+      ContentSource contents = request.contents();
+      if (contents == null) throw new NoDataFoundException("Schema data not sent in the request");
+
+      AdvancedCache<Object, Object> cache = invocationHelper.getRestCacheManager()
+            .getCache(ProtobufMetadataManager.PROTOBUF_METADATA_CACHE_NAME, request);
+
+      NettyRestResponse.Builder builder = new NettyRestResponse.Builder();
+
+      CompletableFuture<Object> putSchema;
+      if (create) {
+         putSchema = cache.putIfAbsentAsync(schemaName, contents.asString()).thenApply(result -> {
+            if (result == null) {
+               builder.status(HttpResponseStatus.CREATED);
+            } else {
+               builder.status(HttpResponseStatus.CONFLICT);
+            }
+            return result;
+         });
+      } else {
+         putSchema = cache.replaceAsync(schemaName, contents.asString()).thenApply(result -> {
+            if (result == null) {
+               builder.status(HttpResponseStatus.NOT_FOUND);
+            } else {
+               builder.status(HttpResponseStatus.OK);
+            }
+            return result;
+         });
+      }
+      return putSchema.thenCompose(r -> {
```

Comment:
This is more like what I would expect. However you didn't fix the indentation on the following lines as they can all now be moved 3 spaces to the left.

Reply:
ok